File: rust/src/bin/003.rs
// Q: What is the largest prime factor of the number 600851475143?
// A: 685

// must install num-integer crate to use this
use num_integer::Roots;

fn is_prime(n: i64) -> bool {
    if n % 2 == 0 { return false }

    for i in (3 .. n.sqrt()).step_by(2) {
        if n % i == 0 { return false }
    }

    true
}

fn main() {
    let num = 600851475143;
    let mut fac: i64 = 1;

    for i in (3 .. num / 2).step_by(2) {
        if is_prime(i) && num % i == 0 {
            fac *= i;
            if fac >= num { break }
        }
    }

    println!("fac => {}", fac);
}

File: golang/003.go
// Q: What is the largest prime factor of the number 600851475143?
// A: 6857

package main

import "fmt"

func main() {
	n := 600851475143
	for i := 2; i < n; i++ {
		if n%i == 0 {
			n = n / i
			i = 1
		}
	}
	fmt.Println(n)
}
